Modules/_abc.c

```diff
@@ -39,50 +39,24 @@ typedef struct {
 static void
 abc_data_dealloc(_abc_data *self)
 {
-    Py_DECREF(self->_abc_registry);
-    Py_DECREF(self->_abc_cache);
-    Py_DECREF(self->_abc_negative_cache);
+    Py_XDECREF(self->_abc_registry);
+    Py_XDECREF(self->_abc_cache);
```

Comment:
I think it is better to do this only for the _registry_.

The point is that many ABCs indeed don't use `.register()`. However, `isinstance()` is called very often in Python (I remember profiling some code and I found that `isinstance()` was the most called function), so that checking every time if caches are already defined may be not worth it.

In other words, using this patter for caches will win some time for startup, but will also loose some time afterwards. @methane Could you please try profiling few programs you have now to see how often `isinstance()` is called?

Reply:
@pppery @ilevkivskyi You're right, but only "negative cache" are used frequently.

I believe single `==NULL` check doesn't have measurable impact.

This is quick benchmark for (1) hitting negative cache and (2) hitting positive cache.
In both cases, set is created and `==NULL` check is just an overhead.

```
# c-abc branch

$ ./python -m perf timeit -s 'import os' -- 'isinstance(42, os._Environ)'
.....................
Mean +- std dev: 502 ns +- 4 ns

$ ./python -m perf timeit -s 'from collections.abc import Sequence; x=[]' -- 'isinstance(x, Sequence)'
.....................
Mean +- std dev: 454 ns +- 3 ns

# c-abc-lazyset branch

$ ./python -m perf timeit -s 'import os' -- 'isinstance(42, os._Environ)'
.....................
Mean +- std dev: 501 ns +- 10 ns

$ ./python -m perf timeit -s 'from collections.abc import Sequence; x=[]' -- 'isinstance(x, Sequence)'
.....................
Mean +- std dev: 449 ns +- 4 ns
```